metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala

```diff
@@ -0,0 +1,263 @@
+   *
+   * errors get logged and sent back as the text contents
+   */
+  def decodedFileContents(uriAsStr: String): Future[String] = {
+
+    val result = Try(URI.create(uriAsStr)) match {
+      case Success(uri) =>
+        if (uri.getScheme != "metalsDecodeFile")
+          Left("URL `$uri` has invalid scheme `${uri.getScheme}`")
+        else {
+          val paramsArr = uri.getQuery.split("&").map(_.split("=", 2))
+          if (paramsArr.exists(_.length != 2))
+            Left(s"URI `$uri` contains invalid query `${uri.getQuery}`")
+          else {
+            Try {
+              paramsArr
+                .map(f =>
+                  URLDecoder.decode(f(0), StandardCharsets.UTF_8) ->
+                    URLDecoder.decode(f(1), StandardCharsets.UTF_8)
+                )
+                .toMap
+            } match {
+              case Success(params) =>
+                if (!params.contains("decoder"))
+                  Left(
+                    s"URI `$uri` contains no `decoder` argument `${uri.getQuery}`"
+                  )
+                else {
+                  if (!params.contains("originalURI"))
+                    Left(
+                      s"URI `$uri` contains no `originalURI` argument `${uri.getQuery}`"
+                    )
+                  else {
+                    val originalURIAsStr = params("originalURI")
+                    Try {
+                      originalURIAsStr.toAbsolutePath
+                    } match {
+                      case Success(filePath) =>
+                        if (!filePath.exists)
+                          Left(s"URI `$uri` file doesn't exist `$filePath`")
+                        else
+                          Right(decodeFileContents(filePath, params))
+                      case Failure(_) =>
+                        Left(s"Bad URI path `$originalURIAsStr`")
+                    }
+                  }
+                }
+              case Failure(_) => Left(s"Bad query parameters `$paramsArr`")
+            }
+          }
+        }
+      case Failure(_) => Left(s"Bad URI `$uriAsStr`")
+    }
+    result match {
+      case Right(success) => success
+      case Left(e) =>
+        scribe.error(e)
+        Future.successful(e)
+    }
+  }
+
+  private def getBoolParam(
+      name: String,
+      params: Map[String, String]
+  ): Boolean = {
+    params.get(name).map(_.toBoolean).getOrElse(false)
+  }
+
+  private def decodeFileContents(
+      filePath: AbsolutePath,
+      params: Map[String, String]
+  ): Future[String] = {
+    params("decoder") match {
+      case "javap" =>
+        decodeJavap(filePath, getBoolParam("verbose", params))
+      case "semanticdb" =>
+        val format = params
+          .get("format")
+          .map(_.toUpperCase)
+          .collect {
+            case "DETAILED" => Format.Detailed
+            case "COMPACT" => Format.Compact
```

Comment:
Since it's fairly simple to add more options I'll take the compact out.  The VSCode file-explorer context menu is already getting busy.  Maybe spin off a Metals sub-menu at one point?

Reply:
Sub menu would be cool probably, we could also just put some as separate commands